activerecord/lib/active_record/associations/collection_proxy.rb

```diff
@@ -1128,7 +1128,18 @@ def pretty_print(pp) # :nodoc:
       %w(insert insert_all insert! insert_all! upsert upsert_all).each do |method|
         class_eval <<~RUBY, __FILE__, __LINE__ + 1
           def #{method}(...)
-            scope.#{method}(...).tap { reset }
+            if loaded? && @association&.target&.any? { |r| r.new_record? }
+              ActiveRecord.deprecator.warn(<<~MSG)
+                Using #{method} on association with unpersisted records
+                (\#{to_a.reject(&:persisted?).map(&:class).uniq.join(', ')})
+                is deprecated. The unpersisted records will be lost after this operation.
+                Please either persist your records first or store them separately before
+                calling #{method}. This will become an error in Rails 8.1.
```

Comment:
I don't think we will ever error. We don't when `update_all` is called with unpersisted records

Reply:
I've cleared out the bit about the error, that is true.